### PR TITLE
#1133: Add support for PFCOUNT.WATCH command

### LINKED ISSUE #1133: Add support for `PFCOUNT.WATCH` command
Add support for reactive queries on the existing `PFCOUNT` command by introducing `PFCOUNT.WATCH`.

This command should send push-responses to subscribed clients whenever the data inside the respective hash set changes.

As a part of this feature, you are also required to add support for the same to the [dicedb-go SDK](https://github.com/DiceDB/dicedb-go) inside the [watch_command.go file](https://github.com/DiceDB/dicedb-go/blob/master/watch_connection.go).

## Reference
1. You may refer to #1047 to understand the changes required to support this command.
2. #924 provides more background on the feature's underlying implementation (but may not be directly relevant to this issue)

## Testing
1. Ensure integration tests are added, you can refer to existing `.WATCH` integration tests.
2. Integration tests should include both resp tests and SDK tests.

### PR DESCRIPTION
**Pull request by iamskp11**
Fixes #1133.
Add support for PFCOUNT.WATCH command, and tests.

### CODE REVIEW

**Comment by iamskp11** [comment]
@JyotinderSingh This is ready for review. Please have a look!

File: integration_tests/commands/resp/pfcountwatch_test.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+	result 		  int64
+}
+
+var PFCountWatchSDKTestCases = []pfcountWatchSDKTestCase{
+	{pfcountWatchKey, "value1", int64(1)},
+	{pfcountWatchKey, "value2", int64(2)},
+	{pfcountWatchKey, "value3", int64(3)},
+	{pfcountWatchKey, "value4", int64(4)},
+}
+
```
Comment (JyotinderSingh) [inline integration_tests/commands/resp/pfcountwatch_test.go:143]: Ideally we should have constants defined separately.
```suggestion
pfCountWatchKey  = "pfcount_watch_key" // assuming this was defined elsewhere
defaultTimeout   = 5 * time.Second
watchFingerprint = "1580567186"
```

File: integration_tests/commands/resp/pfcountwatch_test.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+	val    any
+	result int64
+}
+
+type pfcountWatchSDKWithPFMergeTestCase struct {
+	dest_key_1    string
+	dest_value_1  []string
+	dest_key_2 	  string
+	dest_value_2  []string
+	result 		  int64
```
Comment (JyotinderSingh) [inline integration_tests/commands/resp/pfcountwatch_test.go:134]: style guide doesn't recommend snake case. use camel case.

File: integration_tests/commands/resp/pfcountwatch_test.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+	{pfcountWatchKey, "value4", int64(4)},
+}
+
+var pfcountWatchSDKhWithPFMergeTestCases = []pfcountWatchSDKWithPFMergeTestCase{
+	{"DEST_KEY_1", []string{"a", "b", "c", "d", "e", "f"}, "DEST_KEY_2", []string{"f", "g"}, int64(11)},
+	{"DEST_KEY_3", []string{"a", "b", "c", "d", "e", "f"}, "DEST_KEY_4", []string{"a1", "b1", "c1", "d1", "e1", "f1"}, int64(17)},
+	{"DEST_KEY_5", []string{"f", "g"}, "DEST_KEY_6", []string{"f", "g"}, int64(17)},
+}
+
+func TestPFCountWATCHWithSDK(t *testing.T) {
```
Comment (JyotinderSingh) [inline integration_tests/commands/resp/pfcountwatch_test.go:150]: Maybe we can rewrite this better as follows:
```go
func TestPFCountWatchWithSDK(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
	defer cancel()

	publisher := getLocalSDK(t)
	subscribers := setupSubscribers(t, 3)
	defer cleanupSubscribers(subscribers)

	// Clean up test key before starting
	err := publisher.Del(ctx, pfCountWatchKey)
	assert.NilError(t, err)

	// Setup watch channels
	channels := setupWatchChannels(t, ctx, subscribers)

	t.Run("Basic PFCount Operations", func(t *testing.T) {
		testPFCountOperations(t, ctx, publisher, channels)
	})

	t.Run("PFCount with Merge Operations", func(t *testing.T) {
		testPFCountMergeOperations(t, ctx, publisher, channels)
	})
}

func setupSubscribers(t *testing.T, count int) []watchSubscriber {
	subscribers := make([]watchSubscriber, count)
	for i := range subscribers {
		subscribers[i].client = getLocalSDK(t)
	}
	return subscribers
}

func cleanupSubscribers(subscribers []watchSubscriber) {
	for _, sub := range subscribers {
		if sub.watch != nil {
			sub.watch.Close()
		}
	}
}

func setupWatchChannels(t *testing.T, ctx context.Context, subscribers []watchSubscriber) []<-chan *dicedb.WatchResult {
	channels := make([]<-chan *dicedb.WatchResult, len(subscribers))
	
	for i, subscriber := range subscribers {
		watch := subscriber.client.WatchConn(ctx)
		assert.Assert(t, watch != nil)
		
		firstMsg, err := watch.Watch(ctx, "PFCOUNT", pfCountWatchKey)
		assert.NilError(t, err)
		assert.Equal(t, firstMsg.Command, "PFCOUNT")
		
		subscribers[i].watch = watch
		channels[i] = watch.Channel()
	}
	
	return channels
}

func testPFCountOperations(t *testing.T, ctx context.Context, publisher *SDK, channels []<-chan *dicedb.WatchResult) {
	for _, tc := range pfCountTestCases {
		err := publisher.PFAdd(ctx, tc.Key, tc.Value)
		assert.NilError(t, err)

		verifyWatchResults(t, channels, tc.Result)
	}
}

func testPFCountMergeOperations(t *testing.T, ctx context.Context, publisher *SDK, channels []<-chan *dicedb.WatchResult) {
	for _, tc := range pfCountMergeTestCases {
		// Clean up previous keys
		err := publisher.Del(ctx, tc.DestKey1, tc.DestKey2)
		assert.NilError(t, err)

		// Add values to both keys
		err = publisher.PFAdd(ctx, tc.DestKey1, tc.DestValue1...)
		assert.NilError(t, err)
		err = publisher.PFAdd(ctx, tc.DestKey2, tc.DestValue2...)
		assert.NilError(t, err)

		// Merge keys
		err = publisher.PFMerge(ctx, pfCountWatchKey, tc.DestKey1, tc.DestKey2)
		assert.NilError(t, err)

		verifyWatchResults(t, channels, tc.Expected)
	}
}

func verifyWatchResults(t *testing.T, channels []<-chan *dicedb.WatchResult, expected int64) {
	for _, channel := range channels {
		select {
		case v := <-channel:
			assert.Equal(t, "PFCOUNT", v.Command)
			assert.Equal(t, watchFingerprint, v.Fingerprint)
			assert.DeepEqual(t, expected, v.Data)
		case <-time.After(defaultTimeout):
			t.Fatal("timeout waiting for watch result")
		}
	}
}
```

File: integration_tests/commands/resp/pfcountwatch_test.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+type pfcountWatchTestCase struct {
+	key    string
+	val    any
+	result int64
+}
+
+type pfcountWatchWithPFMergeTestCase struct {
+	dest_key_1    string
+	dest_value_1  []string
```
Comment (JyotinderSingh) [inline integration_tests/commands/resp/pfcountwatch_test.go:23]: use camel case and not snake case for field names

File: integration_tests/commands/resp/pfcountwatch_test.go [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+	dest_key_2 	  string
+	dest_value_2  []string
+	result 		  int64
+}
+
+const (
+	pfcountWatchKey   = "hllkey"
+	pfcountWatchQuery = "PFCOUNT.WATCH %s"
+)
```
Comment (JyotinderSingh) [inline integration_tests/commands/resp/pfcountwatch_test.go:32]: use this to declare the expected fingerprints as well (or define them in a local variable inside the test methods)

**Review by JyotinderSingh — CHANGES_REQUESTED** [review]
Thanks for adding support for PFCOUNT.WATCH command, have left a few comments.

File: integration_tests/commands/resp/pfcountwatch_test.go (hunk 1 above)
Comment (iamskp11) [reply integration_tests/commands/resp/pfcountwatch_test.go:178]: Have defined constants differently for SDK test.

File: integration_tests/commands/resp/pfcountwatch_test.go (hunk 2 above)
Comment (iamskp11) [reply integration_tests/commands/resp/pfcountwatch_test.go]: Changed to camel case.

File: integration_tests/commands/resp/pfcountwatch_test.go (hunk 3 above)
Comment (iamskp11) [reply integration_tests/commands/resp/pfcountwatch_test.go:185]: Have split up the methods

File: integration_tests/commands/resp/pfcountwatch_test.go (hunk 4 above)
Comment (iamskp11) [reply integration_tests/commands/resp/pfcountwatch_test.go]: This is done.

**Comment by iamskp11** [comment]
Have done all requested changes. Please have a look and review @JyotinderSingh

**Review by JyotinderSingh — CHANGES_REQUESTED** [review]
Thanks for addressing the reviews @iamskp11! The changes look good and well-tested. Approved.

**Review by JyotinderSingh — APPROVED** [review]
approved.

**Pull request merged**